### PR TITLE
add_process_metadata: container id: Do not cache whole cgroup file for each process

### PR DESCRIPTION
**Pull request by jtinkus**
- Enhancement
## What does this PR do?

Current solution caches whole content of cgroup file for each process. Let's cache only found container ids. 

Also, current solution skips kube-proxy container processes, because cgroup path has suffix /kube-proxy. 
## Why is it important?

Caching such a pile of useless data has performance impact.
## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- 
Relates #15947

### CODE REVIEW

**Comment by jtinkus** [comment]
Hello! @exekias , @BrookHF, @danmx 
Minor improvements for container id functionality introduced with #15947 I'd like to discuss with you guys.

Firs one is that current solution in master caches huge amount on useless data - whole cgroup file content for all processes. Also it always processes that cached cgroup file content again and again to ask cid even if data is cached. Here i changed cache just to cache container id.

Second issue i discovered was that some kubernetes cgroup paths have suffix after cid. One i found was kube-proxy. For now i hardcoded it there and it works. My question is do you happen to know any other corner cases like that? Maybe I should add suffix also configurable....

With best regards!

**Comment by exekias** [comment]
Thank you for opening this!

> Second issue i discovered was that some kubernetes cgroup paths have suffix after cid. One i found was kube-proxy. For now i hardcoded it there and it works. My question is do you happen to know any other corner cases like that? Maybe I should add suffix also configurable....

Do you know what's the general rule, or why that one behaves differenty? I'm guessing it's because it's an static pod, managed in a different way. We should find the rule for these and include a generic solution

**Comment by jtinkus** [comment]
Completely agree with generic solution and that's what i'm planning to do. 

Finding out the rule needs further investigations, that's why I opened discussion here too, maybe some of you knows about it.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 1]
```diff
@@ -104,7 +107,7 @@ func (p gosigarCidProvider) getCid(cgroups map[string]string) string {
 	for _, path := range cgroups {
 		for _, prefix := range p.cgroupPrefixes {
 			if strings.HasPrefix(path, prefix) {
-				return filepath.Base(path)
+				return filepath.Base(strings.TrimSuffix(path, "/kube-proxy"))
```
Comment (exekias) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go]: I propose we remove this to split concerns in different PRs to make the other part move forward.

**Comment by exekias** [comment]
jenkins test this please

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 1 above)
Comment (jtinkus) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go]: Do you need me to do something for that?

Comment (jtinkus) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go]: in my env i went through all cgroups, there is only this same kube-proxy that differs

Comment (exekias) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go]: yes, could you please add a commit undoing the change to this line? Or just accept this suggestion :innocent: :
```suggestion
				return filepath.Base(path)
```

**Comment by exekias** [comment]
ok to test

**Comment by jtinkus** [comment]
@exekias until i do not know exact rule for cgroup format (and it might not be written anywhere in balck and white, so we could search it forever :)), i see following options:
1) assuming, that differences are in suffix part only and adding configurable suffix list too
2) adding configurable alternative to use regex, something like in my matcher PR #16549

What do you think?

**Comment by exekias** [comment]
I'm ok with adding custom patterns to the `getCid` function. If no pattern is given (none by default) it would use the current approach for automatically guessing. Patterns can be provided to override that behavior.

**Comment by jtinkus** [comment]
Ok i'm happy to add regex pattern as configurable alternative in separate PR.
But still with default we know that kube-proxy processes will have "kube-proxy" added as container id to event, and leaving it like this is behaviour already more like a bug not expected feature :)

**Comment by exekias** [comment]
Yes, I really expect we can investigate this a little more and come up with the generic solution :smile: 

@ChrsMark do you know how the kube-proxy cgroup name is generated by a chance? We can research this a little bit

**Comment by jtinkus** [comment]
@exekias is there anything else i need to do for completing this PR? Or from now on all automatic?

**Comment by ChrsMark** [comment]
> Yes, I really expect we can investigate this a little more and come up with the generic solution 😄
> 
> @ChrsMark do you know how the kube-proxy cgroup name is generated by a chance? We can research this a little bit

@exekias I will check it
